### PR TITLE
LTI 1.3 authentication testing

### PR DESCRIPTION
**Pull request by nwalters512**
This PR implements end-to-end testing of LTI 1.3 authentication being introduced in #8855. It also happens to test the creation and configuration of LTI 1.3 instances.

### CODE REVIEW

File: apps/prairielearn/src/tests/lti13.test.ts [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+      .setExpirationTime('1h')
+      .setSubject('a555090c-8355-4b58-b315-247612cc22f0')
+      .setAudience(CLIENT_ID)
+      .sign(joseKey);
+
+    // Run a server to respond to JWKS requests.
+    const app = express();
+    app.get('/jwks', (req, res) => {
+      res.setHeader('Content-Type', 'application/json');
+      res.end(JSON.stringify(keystore.toJSON()));
```
Comment (trombonekenny) [inline apps/prairielearn/src/tests/lti13.test.ts:225]: I'd probably put an explicit `keystore.toJSON(false)` here, just to be safe.

File: apps/prairielearn/src/ee/auth/lti13/lti13Auth.ts [hunk 2]
```diff
@@ -143,8 +143,8 @@ const LTI13Schema = z.object({
   'https://purl.imsglobal.org/spec/lti/claim/target_link_uri': z.string(),
   'https://purl.imsglobal.org/spec/lti/claim/resource_link': z.object({
     id: z.string(),
-    description: z.string().nullable(),
-    title: z.string().nullable(),
+    description: z.string().nullable().optional(),
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/src/ee/auth/lti13/lti13Auth.ts:146]: Here are some concrete examples that hopefully help explain this. 

--

An object _must_ have a property `foo`, which can be either a string or `null`:

```ts
z.object({
  foo: z.string().nullable(),
});
```

Equivalent to the following TypeScript type:

```ts
type FooObj = {
  foo: string | null
}
```

---

An object _may_ have a property `foo`; if it has that property, that property must be a string:

```ts
z.object({
  foo: z.string().optional(),
});
```

Equivalent to the following TypeScript type:

```ts
type FooObj = {
  foo?: string
}
```

---

An object _may_ have a property `foo`; if it has that property, that property must be either a string or `null`:

```ts
z.object({
  foo: z.string().nullable().optional(),
});
```

Equivalent to the following TypeScript type:

```ts
type FooObj = {
  foo?: string | null
}
```

---

So, when the spec says a value is optional, you probably want `.nullable().optional()` to allow the value to either be `null` or be omitted entirely. Given that, do you think there are other properties here that should have their schema changed?

File: apps/prairielearn/src/tests/lti13.test.ts [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    // a cookie-aware version of fetch.
+    const fetchWithCookies = fetchCookie(fetchCheerio);
+
+    const startLoginResponse = await fetchWithCookies(`${siteUrl}/pl/lti13_instance/1/auth/login`, {
+      method: 'POST',
+      body: new URLSearchParams({
+        iss: siteUrl,
+        // TODO: what should this be?
+        login_hint: 'login_hint',
+        target_link_uri: `${siteUrl}//pl/lti13_instance/1/course_navigation`,
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/src/tests/lti13.test.ts:150]: In the example `/login`/`/callback` requests you gave me, `target_link_uri` is `https://pl.tbkenny.com/pl/lti13_instance/1/course_navigation`, so I figured I'm mimic that. Do you know how an LMS becomes aware of this value? Is this something we have to configure in each Canvas instance?

File: apps/prairielearn/src/tests/lti13.test.ts [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    // `openid-client` relies on the session to store state, so we need to use
+    // a cookie-aware version of fetch.
+    const fetchWithCookies = fetchCookie(fetchCheerio);
+
+    const startLoginResponse = await fetchWithCookies(`${siteUrl}/pl/lti13_instance/1/auth/login`, {
+      method: 'POST',
+      body: new URLSearchParams({
+        iss: siteUrl,
+        // TODO: what should this be?
+        login_hint: 'login_hint',
```
Comment (nwalters512) [inline apps/prairielearn/src/tests/lti13.test.ts:149]: Per the [IMS Global Security Framework](https://www.imsglobal.org/spec/security/v1p0/), `login_hint` is mandatory and is defined like this:

> Hint to the Authorization Server about the login identifier the End-User might use to log in. The permitted values will be defined in the host specification.

I wish the spec said more about this. In your example login request/response flow, `login_hint` corresponds to `"https://purl.imsglobal.org/spec/lti/claim/lti11_legacy_user_id"` and `"https://purl.imsglobal.org/spec/lti/claim/lti1p1"["user_id"]`. Reading Canvas's docs (https://canvas.instructure.com/doc/api/file.lti_dev_key_config.html), they say that this value just has to be passed *back* to Canvas unchanged. Is there anything else you know about this value that could be relevant?

I'm thinking I should test that this value is present in the redirect to the provider, for completeness.

File: apps/prairielearn/src/ee/auth/lti13/lti13Auth.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/ee/auth/lti13/lti13Auth.ts:146]: I actually think we want `optional` more than `nullable` except in a few cases where the attribute gets passed but is empty. I don't think that's common in the LTI spec but we can play with it.

I also feel like if we have `z.any().nullable().optional()` we can probably remove it. I added in those things before I changed it to `passthrough()` so I was worried about loose checking. Now I think we could pull some of that schema back just to things that we 1) know are required per the spec and 2) actually use/parse/interact with. 

I have that perspective today after thinking about your testing environment.... like what would we want to put into the faked LTI 1.3 claims to test on the other side.

[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/src/ee/auth/lti13/lti13Auth.ts:146]: @jonatanschroeder TIL about `.nullish()`, that's super useful and definitely a better pattern.

File: apps/prairielearn/src/tests/lti13.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/src/tests/lti13.test.ts:150]: Good call that we should also test that user information was updated correctly. I'll work on that.

[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/tests/lti13.test.ts:150]: > In the example `/login`/`/callback` requests you gave me, `target_link_uri` is `https://pl.tbkenny.com/pl/lti13_instance/1/course_navigation`, so I figured I'm mimic that. Do you know how an LMS becomes aware of this value? Is this something we have to configure in each Canvas instance?

The pl/lti13_instance/1/config page is the way PL can tell the LMS how to configure itself for PL. That's designed to be computer and human parseable. With Canvas, even if you consume the JSON, you can edit it after the fact but it does some of the heavy lifting.

File: apps/prairielearn/src/tests/lti13.test.ts (hunk 1 above)
Comment (nwalters512) [reply apps/prairielearn/src/tests/lti13.test.ts]: Done in 0d88dc3.

File: apps/prairielearn/src/tests/lti13.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/src/tests/lti13.test.ts:151]: I added tests for both `users` and `lti13_users` in 0d88dc3.

**Pull request merged**